File: unity/Rotgrid/Assets/Scripts/Systems/Shop.cs
using System.Collections.Generic;
using UnityEngine;

namespace Rotgrid
{
    /// <summary>A wall-mounted weapon: buy it once, then buy ammo for it.</summary>
    public class WallBuy : Interactable
    {
        public readonly WeaponDef def;
        readonly MachineInfo _info;

        public WallBuy(WallBuyDef spot, Transform parent)
            : base(spot.id, new Vector3(spot.x, spot.y + 1.2f, spot.z), 2.5f, 1)
        {
            def = Weapons.All[spot.weapon];
            var go = Props.WallBuy(def, parent);
            go.transform.position = new Vector3(spot.x, spot.y + 1.5f, spot.z);
            go.transform.rotation = Quaternion.Euler(0, spot.rot, 0);
            _info = go.GetComponent<MachineInfo>();
        }

        public override Prompt GetPrompt(GameRun run)
        {
            if (run.weapons.Owns(def.id))
            {
                if (run.weapons.AmmoFull(def.id)) return Prompt.Blocked(def.name, "AMMO FULL");
                return Prompt.Buy(def.name + " AMMO", def.wallAmmoCost, "");
            }
            if (def.wallCost <= 0) return Prompt.Blocked(def.name, "STANDARD ISSUE");
            return Prompt.Buy("BUY " + def.name, def.wallCost, def.ClassLabel);
        }

        public override bool Activate(GameRun run)
        {
            if (run.weapons.Owns(def.id))
            {
                if (run.weapons.AmmoFull(def.id)) return false;
                if (!run.player.Spend(def.wallAmmoCost)) { run.Deny(); return false; }
                run.weapons.RefillWeapon(def.id);
                Game.Audio.Play(Sfx.UiSelect, 1f);
                run.Toast("AMMO PURCHASED", ToastKind.Good, false);
                return true;
            }
            if (def.wallCost <= 0) return false;
            if (!run.player.Spend(def.wallCost)) { run.Deny(); return false; }
            run.weapons.GiveWeapon(def);
            Game.Audio.Play(Sfx.UiSelect, 1f);
            return true;
        }

        public void Tick(float time)
        {
            if (_info != null && _info.spinTarget != null)
            {
                var p = _info.spinTarget.localPosition;
                _info.spinTarget.localPosition = new Vector3(p.x, 0.09f + Mathf.Sin(time * 1.6f + pos.x) * 0.02f, p.z);
            }
        }
    }

    /// <summary>Armour plating dispenser.</summary>
    public class ArmorStation : Interactable
    {
        readonly int _cost;

        public ArmorStation(ArmorDef spot, Transform parent)
            : base(spot.id, new Vector3(spot.x, spot.y + 1.1f, spot.z), 2.3f, 1)
        {
            _cost = spot.cost;
            var g = new GameObject("armor_" + spot.id);
            g.transform.SetParent(parent, false);
            g.transform.position = new Vector3(spot.x, spot.y, spot.z);
            g.transform.rotation = Quaternion.Euler(0, spot.rot, 0);
            Props.Box(1.2f, 1.8f, 0.35f, Props.DarkMetal, new Vector3(0, 0.9f, 0), g.transform);
            Props.Box(1.0f, 0.28f, 0.16f, Art.Emissive("em_armor", Util.Hex(0x4aa3ff), 1.6f),
                new Vector3(0, 1.5f, 0.2f), g.transform);
            for (int i = 0; i < 3; i++)
                Props.Box(0.72f, 0.1f, 0.3f, Props.Metal, new Vector3(0, 0.45f + i * 0.32f, 0.14f), g.transform);
            var sign = Props.Quad(1.1f, 0.4f, Art.Emissive("sign_armor", Color.white, 1.2f,
                PixelFont.Sign("PLATING", spot.cost.ToString(), Util.Hex(0x4aa3ff), Util.Hex(0x080b0d), 256, 96)),
                new Vector3(0, 1.85f, 0.2f), g.transform);
            sign.transform.localRotation = Quaternion.Euler(0, 180f, 0);
        }

        public override Prompt GetPrompt(GameRun run)
        {
            if (run.player.armor >= 100f) return Prompt.Blocked("ARMOUR PLATING", "PLATES FULL");
            return Prompt.Buy("ARMOUR PLATING", _cost, "ABSORBS 65% OF DAMAGE");
        }

        public override bool Activate(GameRun run)
        {
            if (run.player.armor >= 100f) return false;
            if (!run.player.Spend(_cost)) { run.Deny(); return false; }
            run.player.AddArmor(50f);
            Game.Audio.Play(Sfx.UiSelect, 1f);
            run.Toast("PLATES APPLIED", ToastKind.Good, false);
            return true;
        }
    }

    public class ShopSystem
    {
        readonly List<WallBuy> _walls = new List<WallBuy>();

        public ShopSystem(InteractionSystem interaction, Transform parent)
        {
            foreach (var spot in MapData.WallBuys)
            {
                var w = new WallBuy(spot, parent);
                interaction.Add(w);
                _walls.Add(w);
            }
            foreach (var spot in MapData.ArmorStations)
                interaction.Add(new ArmorStation(spot, parent));
        }

        public void Update(float time)
        {
            foreach (var w in _walls) w.Tick(time);
        }
    }
}

File: src/systems/Shop.js
import * as THREE from 'three';
import { makeWallBuy, box, Mats } from '../world/Props.js';
import { Tex } from '../world/Textures.js';
import { Interactable } from './Interaction.js';
import { MAP } from '../world/MapData.js';
import { WEAPONS } from '../weapons/WeaponDefs.js';
import { Audio } from '../core/AudioEngine.js';

class WallBuy extends Interactable {
  constructor(spot, scene) {
    const def = WEAPONS[spot.weapon];
    super({ id: spot.id, x: spot.x, y: spot.y + 1.2, z: spot.z, radius: 2.5, priority: 1 });
    this.def = def;
    this.group = makeWallBuy(def);
    this.group.position.set(spot.x, spot.y + 1.5, spot.z);
    this.group.rotation.y = spot.rot;
    scene.add(this.group);
  }

  prompt(ctx) {
    const owned = ctx.ownsWeapon(this.def.id);
    if (owned) {
      const full = ctx.weaponAmmoFull(this.def.id);
      if (full) return { text: this.def.name, sub: 'AMMO FULL', blocked: true };
      return { text: `${this.def.name} AMMO`, cost: this.def.wallAmmoCost };
    }
    if (!this.def.wallCost) return { text: this.def.name, sub: 'STANDARD ISSUE', blocked: true };
    return { text: `BUY ${this.def.name}`, cost: this.def.wallCost, sub: this.def.category.toUpperCase() };
  }

  activate(ctx) {
    const owned = ctx.ownsWeapon(this.def.id);
    if (owned) {
      if (ctx.weaponAmmoFull(this.def.id)) return false;
      if (!ctx.player.spend(this.def.wallAmmoCost)) { ctx.deny(); return false; }
      ctx.refillWeapon(this.def.id);
      Audio.play('ui', { kind: 'select' });
      ctx.toast('AMMO PURCHASED', 'pts');
      return true;
    }
    if (!this.def.wallCost) return false;
    if (!ctx.player.spend(this.def.wallCost)) { ctx.deny(); return false; }
    ctx.giveWeapon(this.def);
    Audio.play('ui', { kind: 'select' });
    return true;
  }

  update(dt, time) {
    const g = this.group.userData.weaponModel;
    if (g) g.position.y = 0.06 + Math.sin(time * 1.6 + this.pos.x) * 0.02;
    void dt;
  }
}

class ArmorStation extends Interactable {
  constructor(spot, scene) {
    super({ id: spot.id, x: spot.x, y: spot.y + 1.1, z: spot.z, radius: 2.3, priority: 1 });
    this.cost = spot.cost;
    this.group = new THREE.Group();
    this.group.add(box(1.2, 1.8, 0.35, Mats.darkMetal, 0, 0.9, 0));
    this.group.add(box(1.0, 0.28, 0.16, Mats.emissive(0x4aa3ff, 1.6), 0, 1.5, 0.2));
    for (let i = 0; i < 3; i++) {
      this.group.add(box(0.72, 0.1, 0.3, Mats.metal, 0, 0.45 + i * 0.32, 0.14));
    }
    const sign = new THREE.Mesh(new THREE.PlaneGeometry(1.1, 0.4), new THREE.MeshBasicMaterial({
      map: Tex.label('PLATING', { fg: '#4aa3ff', sub: `${spot.cost}` }),
    }));
    sign.position.set(0, 1.85, 0.2);
    this.group.add(sign);
    this.group.position.set(spot.x, spot.y, spot.z);
    this.group.rotation.y = spot.rot;
    scene.add(this.group);
  }

  prompt(ctx) {
    if (ctx.player.armor >= 100) return { text: 'ARMOUR PLATING', sub: 'PLATES FULL', blocked: true };
    return { text: 'ARMOUR PLATING', cost: this.cost, sub: 'ABSORBS 65% OF DAMAGE' };
  }

  activate(ctx) {
    if (ctx.player.armor >= 100) return false;
    if (!ctx.player.spend(this.cost)) { ctx.deny(); return false; }
    ctx.player.addArmor(50);
    Audio.play('ui', { kind: 'select' });
    ctx.toast('PLATES APPLIED', 'pts');
    return true;
  }

  update() {}
}

/** Wall-mounted purchases: weapons, ammo, and armour plating. */
export class ShopSystem {
  constructor(world, scene, interaction) {
    this.stations = [];
    for (const spot of MAP.wallBuys) {
      const w = new WallBuy(spot, scene);
      interaction.add(w);
      this.stations.push(w);
    }
    for (const spot of MAP.armorStations) {
      const a = new ArmorStation(spot, scene);
      interaction.add(a);
      this.stations.push(a);
    }
    void world;
  }

  update(dt, time) {
    for (const s of this.stations) s.update(dt, time);
  }
}
